ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -966,46 +1098,88 @@ private void handleMetadataBlob(BlobData blobData) throws IOException, MessageFo
+                blobInfoFromServer != null ? ByteBuffer.wrap(blobInfoFromServer.getUserMetadata()) : null,
+                cryptoService, kms, (DecryptJob.DecryptJobResult result, Exception exception) -> {
+              logger.trace("Handling decrypt job call back for Metadata chunk {} to set decrypt callback results",
+                  blobId);
+              decryptCallbackResult = result;
+              decryptCallbackException = exception;
+              decryptCallbackResultAvailable = true;
+            }));
+          } else {
+            blobContentAvailableToProcess = true;
+          }
         }
       }
     }
 
     /**
      * Process a simple blob and extract the requested data from the blob.
      * @param blobData the simple blob's data
+     * @param encryptionKey encryption key for the blob. Could be null for non encrypted blob.
      */
-    private void handleSimpleBlob(BlobData blobData) {
+    private void handleSimpleBlob(BlobData blobData, ByteBuffer encryptionKey) {
       totalSize = blobData.getSize();
       chunkSize = totalSize;
       boolean rangeResolutionFailure = false;
       try {
         if (options != null && options.getBlobOptions.getRange() != null) {
-          resolvedByteRange = options.getBlobOptions.getRange().toResolvedByteRange(totalSize);
+          resolvedByteRange =
+              options.getBlobOptions.getRange().toResolvedByteRange(blobInfo.getBlobProperties().getBlobSize());
         }
       } catch (IllegalArgumentException e) {
         onInvalidRange(e);
         rangeResolutionFailure = true;
+        blobContentAvailableToProcess = true;
       }
       if (!rangeResolutionFailure) {
         chunkIdIterator = null;
         dataChunks = null;
         chunkIndex = 0;
         numChunksTotal = 1;
-        chunkIndexToBuffer.put(0, filterChunkToRange(blobData));
-        numChunksRetrieved = 1;
+        if (encryptionKey == null) {
+          chunkIndexToBuffer.put(0, filterChunkToRange(blobData.getStream().getByteBuffer()));
+          numChunksRetrieved = 1;
+          blobContentAvailableToProcess = true;
+        } else {
+          state = ChunkState.Decrypting;
+          blobInfoFromServer = blobInfo;
+          blobInfo = null;
+          logger.trace("Marking the chunk's state to {} and submitting decrypt job for simple blob {}",
+              ChunkState.Decrypting, blobId);
+          exec.submitJob(new DecryptJob(blobId, encryptionKey, blobData.getStream().getByteBuffer(),
+              blobInfoFromServer != null ? ByteBuffer.wrap(blobInfoFromServer.getUserMetadata()) : null, cryptoService,
+              kms, (DecryptJob.DecryptJobResult result, Exception exception) -> {
+            logger.trace("Handling decrypt job call back for simple blob {} to set decrypt callback results", blobId);
+            decryptCallbackResult = result;
+            decryptCallbackException = exception;
+            decryptCallbackResultAvailable = true;
```

Comment:
Should the encryption callbacks generally be calling `RouterCallback.onPollReady()` to wake up the selector? Otherwise, the router could be waiting at `networkClient.sendAndPoll()` when there is useful work to do. @pnarayanan

Reply:
Yes, I did think about it and even had it in my patch at some point while coding. But felt the NonBlockingRouter will be doing a infinite loop of poll and hence removed it. Lets see what priyesh has to say.

Reply:
I have fixed this.